index.d.ts

```diff
@@ -1,10 +1,15 @@
 declare function deepmerge<T>(x: Partial<T>, y: Partial<T>, options?: deepmerge.Options): T;
 declare function deepmerge<T1, T2>(x: Partial<T1>, y: Partial<T2>, options?: deepmerge.Options): T1 & T2;
 
+declare interface CustomMerge<T1, T2> {
+  (x: Partial<T1>, y: Partial<T1 | T2>): T1 | T1 & T2;
+}
+
 declare namespace deepmerge {
 	export interface Options {
 		arrayMerge?(target: any[], source: any[], options?: Options): any[];
 		clone?: boolean;
+		customMerge?: (key: string) => <T1, T2>(x: Partial<T1>, y: Partial<T1 | T2>) => T1 | T1 & T2;
```

Comment:
There should be a corresponding test line or two in `test/typescript.ts` that fails without this addition.

Reply:
I'm no longer sure that this is the best way to type `customMerge`. This would ensure that the type returned is either the type that was passed, or an intersection of both types passed. It seems that it'd be preferable to allow any type to be returned from a custom merge function, as there may be use cases where users want to return something other than the types passed in.

Does that make sense? Not sure if I'm being clear there

Reply:
I wrote a few tests to see what works/fails with the typing, the standard merge functions we've used with this work, here's an example of something which wouldn't:

```Typescript
const options4: merge.Options = {
    customMerge: (key, options) => {
        return false;
    }
};
```
This cause the typescript test to output an error. This however works fine, so it ought to still allow for situations where certain keys do not return a function:
```typescript
const options3: merge.Options = {
    customMerge: (key) => {
        if (key === 'foo') {
          return (target, source) => target + source;
        }
    }
}
```

It seems to me that the flexibility inherent in the problem of allowing custom merges means that `any` is probably the best that can be done. It may be that others have a better suggestion for this, though, and I'd be interested to hear anything.

Reply:
Any further thoughts on this? My thoughts on this is that it would always need to be typed as any, as there's no way to ascertain the return type for any given key without first running the function to generate it.

Reply:
I'd be okay with `any`, but I think the current way is fine, where it's typed as an optional function.  People can `ts-ignore` if they want to return something other than `undefined`.

Reply:
I think we're actually along the same lines with this, what I was suggesting was:
```typescript
customMerge?: (key: string) => (x: any, y: any) => any;
```
That way, custom merge needs to be a function which takes in a string, but can return any function which takes two arguments and returns any value.

Reply:
ah, I see - I think the definition of that second function is right, but the first function needs to be changed, since it's acceptable for the `customMerge` function to not return a function.

Reply:
Thanks, that makes sense; I'll have a look at this later on and see if I can get type that looks good

Reply:
I've updated this now, hopefully it represents what we've been discussing, but if there are any further changes needed just let me know